File: core/src/main/java/org/mifos/mobilewallet/core/base/UseCaseThreadPoolScheduler.java
package org.mifos.mobilewallet.core.base;

import android.os.Handler;

import java.util.concurrent.Executors;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

/**
 * Executes asynchronous tasks using a {@link ThreadPoolExecutor}.
 * <p>
 * See also {@link Executors} for a list of factory methods to create common
 * {@link java.util.concurrent.ExecutorService}s for different scenarios.
 */
public class UseCaseThreadPoolScheduler implements UseCaseScheduler {

    public static final int POOL_SIZE = 2;
    public static final int MAX_POOL_SIZE = 40;
    public static final int TIMEOUT = 60;
    private final Handler mHandler = new Handler();
    ThreadPoolExecutor mThreadPoolExecutor;

    public UseCaseThreadPoolScheduler() {
        mThreadPoolExecutor = new ThreadPoolExecutor(POOL_SIZE, MAX_POOL_SIZE, TIMEOUT,
                TimeUnit.SECONDS, new ThreadPoolQueue(MAX_POOL_SIZE));
    }

    @Override
    public void execute(Runnable runnable) {
        mThreadPoolExecutor.execute(runnable);
    }


    @Override
    public <V extends UseCase.ResponseValue> void notifyResponse(final V response,
            final UseCase.UseCaseCallback<V>
                    useCaseCallback) {

        mHandler.post(new Runnable() {
            @Override
            public void run() {
                useCaseCallback.onSuccess(response);
            }
        });
    }

    @Override
    public <V extends UseCase.ResponseValue> void onError(final String message,
            final UseCase.UseCaseCallback<V> useCaseCallback) {

        mHandler.post(new Runnable() {
            @Override
            public void run() {
                useCaseCallback.onError(message);
            }
        });
    }

}

File: core/data/src/main/java/org/mifospay/core/data/base/UseCaseThreadPoolScheduler.kt
/*
 * Copyright 2024 Mifos Initiative
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * See https://github.com/openMF/mobile-wallet/blob/master/LICENSE.md
 */
package org.mifospay.core.data.base

import android.os.Handler
import android.os.Looper
import org.mifospay.core.data.base.UseCase.UseCaseCallback
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

/**
 * Executes asynchronous tasks using a [ThreadPoolExecutor].
 *
 *
 * See also [ThreadPoolExecutor] for a list of factory methods to create common
 * [java.util.concurrent.ExecutorService]s for different scenarios.
 */
class UseCaseThreadPoolScheduler : UseCaseScheduler {

    private val mHandler = Handler(Looper.getMainLooper())
    private var mThreadPoolExecutor: ThreadPoolExecutor = ThreadPoolExecutor(
        POOL_SIZE,
        MAX_POOL_SIZE,
        TIMEOUT.toLong(),
        TimeUnit.SECONDS,
        ThreadPoolQueue(MAX_POOL_SIZE),
    )

    override fun execute(runnable: Runnable?) {
        mThreadPoolExecutor.execute(runnable)
    }

    override fun <V : UseCase.ResponseValue?> notifyResponse(
        response: V,
        useCaseCallback: UseCaseCallback<V>?,
    ) {
        mHandler.post { useCaseCallback!!.onSuccess(response) }
    }

    override fun <V : UseCase.ResponseValue?> onError(
        message: String?,
        useCaseCallback: UseCaseCallback<V>?,
    ) {
        mHandler.post { useCaseCallback!!.onError(message!!) }
    }

    companion object {
        const val POOL_SIZE = 2
        const val MAX_POOL_SIZE = 40
        const val TIMEOUT = 60
    }
}
